File: typescript/src/api/mcp-client.ts
// JSON-RPC 2.0 client for MCP tool calls (tools/call method).

let nextId = 1

export interface JsonRpcRequest {
  jsonrpc: '2.0'
  id: number
  method: string
  params: Record<string, unknown>
}

export interface JsonRpcResponse<T = unknown> {
  jsonrpc: '2.0'
  id: number
  result?: T
  error?: {
    code: number
    message: string
    data?: unknown
  }
}

interface ToolResultContent {
  type: string
  text?: string
}

interface ToolResult {
  content: ToolResultContent[]
  isError?: boolean
}

export class McpError extends Error {
  code: number
  data?: unknown

  constructor(code: number, message: string, data?: unknown) {
    super(message)
    this.name = 'McpError'
    this.code = code
    this.data = data
  }
}

export function buildJsonRpcRequest(method: string, params: Record<string, unknown> = {}): JsonRpcRequest {
  return {
    jsonrpc: '2.0',
    id: nextId++,
    method,
    params,
  }
}

export function resetRequestId(): void {
  nextId = 1
}

/**
 * Call an MCP tool via JSON-RPC tools/call and unwrap the ToolResult.
 *
 * Sends: { method: "tools/call", params: { name: toolName, arguments: params } }
 * Receives: { result: { content: [{ type: "text", text: "<json>" }] } }
 * Returns: parsed JSON from the first text content block.
 */
export async function mcpCall<T>(
  endpoint: string,
  toolName: string,
  params: Record<string, unknown> = {},
  agentId?: string,
): Promise<T> {
  const body = buildJsonRpcRequest('tools/call', {
    name: toolName,
    arguments: params,
  })

  const headers: Record<string, string> = { 'Content-Type': 'application/json' }
  if (agentId) {
    headers['X-Agent-ID'] = agentId
  }

  const res = await fetch(endpoint, {
    method: 'POST',
    headers,
    body: JSON.stringify(body),
  })

  if (!res.ok) {
    throw new McpError(-1, `HTTP ${res.status}: ${res.statusText}`)
  }

  const json = (await res.json()) as JsonRpcResponse<ToolResult>

  if (json.error) {
    throw new McpError(json.error.code, json.error.message, json.error.data)
  }

  if (json.result === undefined) {
    throw new McpError(-1, 'Server returned success response with no result field')
  }

  const toolResult = json.result

  if (toolResult.isError) {
    const errorText = toolResult.content?.[0]?.text ?? 'Tool call failed'
    throw new McpError(-32000, errorText)
  }

  const textContent = toolResult.content?.find(c => c.type === 'text')
  if (!textContent?.text) {
    throw new McpError(-1, 'Tool returned no text content')
  }

  return JSON.parse(textContent.text) as T
}

File: java/src/main/java/com/trading/mcp/McpClient.java
package com.trading.mcp;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.trading.config.AgentProperties;
import com.trading.mcp.dto.McpResponse;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;
import org.springframework.web.client.RestClient;

import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * JSON-RPC 2.0 client for MCP servers.
 * Uses Spring RestClient (sync) — safe with Java 21 virtual threads.
 * All MCP servers share the same protocol; only the URL and tool name differ.
 */
@Component
public class McpClient {

    private static final Logger log = LoggerFactory.getLogger(McpClient.class);

    private final RestClient restClient;
    private final AgentProperties props;
    private final ObjectMapper mapper;
    private final AtomicInteger seq = new AtomicInteger(1);

    public McpClient(AgentProperties props, ObjectMapper mapper) {
        this.props = props;
        this.mapper = mapper;
        this.restClient = RestClient.builder()
                .defaultHeader("Content-Type", "application/json")
                .defaultHeader("X-Agent-ID", props.id())
                .defaultHeader("X-Api-Key", props.apiKey())
                .build();
    }

    /**
     * Call a tool on an MCP server and deserialize the text content into responseType.
     *
     * @param mcpUrl       target MCP endpoint (e.g. http://swiftward-server:8095/mcp/market)
     * @param toolName     namespaced tool name (e.g. "market/get_candles")
     * @param arguments    tool arguments (serialized as JSON object)
     * @param responseType class to deserialize the text content into
     */
    public <T> T call(String mcpUrl, String toolName, Object arguments, Class<T> responseType) {
        var body = Map.of(
                "jsonrpc", "2.0",
                "id", seq.getAndIncrement(),
                "method", "tools/call",
                "params", Map.of("name", toolName, "arguments", arguments)
        );

        log.debug("MCP call: {} -> {}", mcpUrl, toolName);

        McpResponse response = restClient.post()
                .uri(mcpUrl)
                .body(body)
                .retrieve()
                .body(McpResponse.class);

        if (response == null) {
            throw new McpException("Null response from " + toolName);
        }
        if (response.error() != null) {
            throw new McpException("MCP error from %s: [%d] %s"
                    .formatted(toolName, response.error().code(), response.error().message()));
        }
        if (response.result() == null
                || response.result().content() == null
                || response.result().content().isEmpty()) {
            throw new McpException("Empty result content from " + toolName);
        }

        String text = response.result().content().get(0).text();
        try {
            return mapper.readValue(text, responseType);
        } catch (JsonProcessingException e) {
            throw new McpException("Failed to parse response from %s: %s".formatted(toolName, e.getMessage()), e);
        }
    }

    /**
     * Convenience overload with a pre-built arguments list for tools that accept list params.
     */
    public <T> T call(String mcpUrl, String toolName, Map<String, Object> arguments, Class<T> responseType) {
        return call(mcpUrl, toolName, (Object) arguments, responseType);
    }
}
